aiosip/client.py

```diff
@@ -0,0 +1,355 @@
+
+        from_details.add_tag()
+
+        if not call_id:
+            call_id = str(uuid.uuid4())
+
+        if not contact_details:
+            host, port = self._addr
+
+            # No way to get the public local addr in UDP. Allow an override or select the From host
+            # Maybe with https://bugs.python.org/issue31203
+            # if self._app.defaults['override_contact_host']:
+            #     host = self._app.defaults['override_contact_host']
+            # elif host == '0.0.0.0' or host.startswith('127.'):
+            #     host = from_details['uri']['host']
+
+            contact_details = Contact(
+                {
+                    'uri': 'sip:{username}@{host_and_port};transport={protocol}'.format(
+                        username=from_details['uri']['user'],
+                        host_and_port=utils.format_host_and_port(host, port),
+                        protocol=self._proto_type.__name__.lower()
+                    )
+                }
+            )
+
+        dialog = dialog_factory(
+            method=method,
+            from_details=from_details,
+            to_details=to_details,
+            contact_details=contact_details,
+            call_id=call_id,
+            peer=self,
+            password=password,
+            headers=headers,
+            payload=payload,
+            cseq=cseq,
+            inbound=inbound,
+            **kwargs
+        )
+
+        LOG.debug('Creating: %s', dialog)
+
+        self._dialogs[dialog.dialog_id] = dialog
+        self._dialogs[frozenset((dialog.original_msg.to_details['params'].get('tag'), None, dialog.call_id))] = dialog
+        return dialog
+
+    async def request(self, method, from_details, to_details, contact_details=None, password=None, call_id=None,
+                      headers=None, cseq=0, payload=None, dialog_factory=Dialog, timeout=None, **kwargs):
+
+        if not self._protocol:
+            await self.connect()
+
+        dialog = self._create_dialog(method=method,
+                                     from_details=from_details,
+                                     to_details=to_details,
+                                     contact_details=contact_details,
+                                     headers=headers,
+                                     payload=payload,
+                                     password=password,
+                                     call_id=call_id,
+                                     cseq=cseq,
+                                     dialog_factory=dialog_factory,
+                                     **kwargs)
+
+        try:
+            await dialog.start(timeout=timeout)
+            return dialog
+        except asyncio.CancelledError:
+            dialog.cancel()
+            raise
+        except exceptions.AuthentificationFailed:
+            await dialog.close(fast=True)
+            raise
+
+    @asynccontextmanager
```

Comment:
For me, yes.
Is it a problem for you ?

Reply:
I used `asynccontextmanager` since the code I was working on is python 3.7 . If required I have no problem dropping 3.6 support, in that particular case I believe there is a backport of `asynccontextmanager` if we decide to go with that.